src/SubHask/Algebra/Matrix.hs

```diff
@@ -0,0 +1,340 @@
+{-# OPTIONS_GHC -fno-warn-name-shadowing #-}
+{-# OPTIONS_GHC -fno-warn-unused-binds #-}
+
+module SubHask.Algebra.Matrix
+    ( Matrix (..)
+    , unsafeToModuleM
+    , ValidMatrix
+    , MatrixField
+    , mmult
+    , transposeM
+    , row
+    , col
+    , (!!)
+    )
+    where
+
+import qualified Prelude as P
+
+import Data.Primitive hiding (sizeOf)
+import GHC.TypeLits
+import Control.Monad.Primitive
+import Control.Monad
+import qualified Data.Primitive as Prim
+import qualified Data.Foldable
+
+import qualified SubHask.Algebra.Vector as V
+import SubHask.Algebra
+import SubHask.Category
+import SubHask.Internal.Prelude
+
+import qualified Numeric.LinearAlgebra
+
+{-# INLINE extendDimensions #-}
+extendDimensions :: Int -> Int
+extendDimensions = roundUpToNearest 4 -- i+4-i`rem`4
+
+type MatrixField r =
+    ( VectorSpace r
+    , Real r
+    , OrdField r
+    , Prim r
+    , Storable r
+    , Numeric.LinearAlgebra.Field r
+    , P.Num (Numeric.LinearAlgebra.Vector r)
+    )
+
+-- | The type of dynamic or statically sized matrices.
+data family Matrix r (m::k) (n::k)
+
+type instance Scalar (Matrix r m n) = Scalar r
+type instance Logic (Matrix r m n) = Logic r
+type instance Matrix r m n >< a = Matrix (r><a) m n 
+
+type ValidMatrix r m n = ( (Matrix r m n><Scalar r)~Matrix r m n, Prim r)
+
+type instance Index (Matrix r m n) = (Int, Int)
+type instance Elem (Matrix r m n) = Scalar r
+type instance SetElem (Matrix r m n) b = Matrix b m n
+
+-- | dynamic matrix type
+data instance Matrix r (m::Symbol) (n::Symbol) = Matrix_Dynamic
+    {-#UNPACK#-}!ByteArray
```

Comment:
Ok.  I can have a crack at doing this.  I assume it's best to stick with Matrix.hs thinking that there's a ghc 8.0 refactor happening at some stage?